vessel-runtime/src/main/java/com/textnow/android/vessel/Profiler.kt

```diff
@@ -0,0 +1,358 @@
+                data = events.map { listOf(it.niceName(), hitCountOf(it)) },
+                summaryRow = listOf("", events.sumOf { hitCountOf(it) })
+            )
+        }
+
+Database I/O times, by thread/coroutine, sorted by time spent
+${
+            table(
+                headings = listOf("type", "name") + spans.map { it.niceName() } + listOf("total"),
+                data = workerData.map {
+                    listOf(
+                        it.worker.type,
+                        it.worker.name
+                    ) + spans.map { span -> it.spans[span]?.totalDurationMS ?: 0 } + listOf(
+                        it.totalDurationMS
+                    )
+                },
+                summaryRow = listOf(
+                    "", /* type */
+                    "", /* name */
+                ) + spans.map { timeIn(it) } + spans.sumOf { timeIn(it) }
+
+            )
+        }
+
+* all times in ms
+"""
+    }
+}
+
+/**
+ * Profiler for use by [Vessel]
+ * Can be used to tracks time spent performing database operations by worker (thread/coroutine), and
+ * to count cache hits
+ */
+internal interface Profiler {
+    /**
+     * Track execution time for [block]
+     * This is added to the running total of spent in the given [span] for the current coroutine
+     */
+    suspend fun <T> time(span: Span, block: suspend () -> T): T
+
+    /**
+     * Track execution time for [block]
+     * This is added to the running total of time spent in the given [span] for the current thread
+     */
+    fun <T> timeBlocking(span: Span, block: () -> T): T
+
+    /**
+     * Increment the hit count for the given event
+     * This is added to the running total of hits for given [event], for the current coroutine
+     */
+    suspend fun count(event: Event)
+
+    /**
+     * Increment the hit count for the given event
+     * This is added to the running total of hits for given [event], for the current thread
+     */
+    fun countBlocking(event: Event)
+
+    /**
+     * Retrieve a copy of the current [ProfileData]
+     * Note that this is a copy and will not reflect future changes
+     */
+    val snapshot: ProfileData
+}
+
+internal class DummyProfiler : Profiler {
+    override suspend fun <T> time(span: Span, block: suspend () -> T): T {
+        return block()
+    }
+
+    override fun <T> timeBlocking(span: Span, block: () -> T): T {
+        return block()
+    }
+
+    override suspend fun count(event: Event) {
+    }
+
+    override fun countBlocking(event: Event) {
+    }
+
+    override val snapshot get() = ProfileData(emptySet())
+
+}
+
+internal class ProfilerImpl : Profiler {
+    private val workers = ConcurrentHashMap<Worker, WorkerData>()
+
+    private fun timeImpl(worker: Worker, span: Span, duration: Long) {
+        val workerData = workers.getOrPut(worker) {
+            WorkerData(worker)
+        }
+
+        // Note:  synchronized is safe - and required - here:
```

Comment:
multiline comments should use the `/** ... */` syntax
https://kotlinlang.org/docs/coding-conventions.html#documentation-comments